modules/Bio/EnsEMBL/Utils/Tree/Interval/Mutable/Node.pm

```diff
@@ -0,0 +1,664 @@
+  if ($self->left and $self->right) {
+    $self->{max} = max $self->left->{max}, $self->right-{max}, $high;
+  } elsif ($self->left and !$self->right) {
+    $self->{max} = max $self->left->{max}, $high;
+  } elsif (!$self->left and $self->right) {
+    $self->{max} = max $self->right->{max}, $high;
+  } else {
+    $self->{max} = $high;
+  }
+
+  $self->parent->_update_parents_max if $self->parent;
+}
+
+=head2 _rebalance
+
+Rebalances the tree if the height value between two nodes of the same parent is greater than two. 
+There are 4 cases that can happen:
+
+Left-Left case:
+         z                                      y
+        / \                                   /   \
+       y   T4      Right Rotate (z)          x     z
+      / \          - - - - - - - - ->       / \   / \
+     x   T3                                T1 T2 T3 T4
+    / \
+  T1   T2
+
+Left-Right case:
+       z                               z                           x
+      / \                             / \                        /   \
+     y   T4  Left Rotate (y)         x  T4  Right Rotate(z)     y     z
+    / \      - - - - - - - - ->     / \      - - - - - - - ->  / \   / \
+  T1   x                           y  T3                      T1 T2 T3 T4
+      / \                         / \
+    T2   T3                      T1 T2 
+
+Right-Right case:
+    z                               y
+   / \                            /   \
+  T1  y     Left Rotate(z)       z     x
+     / \   - - - - - - - ->     / \   / \
+    T2  x                      T1 T2 T3 T4
+       / \
+      T3 T4
+
+Right-Left case:
+     z                            z                            x
+    / \                          / \                         /   \
+   T1  y   Right Rotate (y)     T1  x      Left Rotate(z)   z     y
+      / \  - - - - - - - - ->      / \   - - - - - - - ->  / \   / \
+     x  T4                        T2  y                   T1 T2 T3 T4
+    / \                              / \
+  T2   T3                           T3 T4
+
+=cut 
+
+sub _rebalance {
+  my $self = shift;
+
+  my ($left, $right) = ($self->left, $self->right);
+  
+  if (_height($left) - _height($right) >= 2) {
+    if (_height($left->left) >= _height($left->right)) {
+      # Left-Left case
+      $self->_right_rotate;
+      $self->_update_max_right_rotate;
+    } else {
+      # Left-Right case
+      $left->_left_rotate;
+      $self->_right_rotate;
+      $self->_update_max_right_rotate;
+    }
+  } elsif (_height($right) - _height($left) >= 2) {
+    if (_height($right->right) >= _height($right->left)) {
+      # Right-Right case
+      $self->_left_rotate;
+      $self->_update_max_left_rotate;
+    } else {
+      # Right-Left case
+      $right->_right_rotate;
+      $self->_left_rotate;
+      $self->_update_max_left_rotate;
+    }
+  }
+}
+
+sub _left_rotate {
+  my $self = shift;
+
+  my $right = $self->right;
+
+  $right->parent($self->parent);
+  unless (defined $right->parent) {
```

Comment:
Could this not be swapped so the `if (defined $right->parent)` then perform what is in the `else` block